diff -ur ueagle-atm4/driver/ueagle-atm.c ueagle-atm4.new/driver/ueagle-atm.c
--- ueagle-atm4/driver/ueagle-atm.c	2006-11-23 21:13:28.000000000 +0100
+++ ueagle-atm4.new/driver/ueagle-atm.c	2007-07-01 13:12:26.042633019 +0200
@@ -1511,7 +1511,7 @@
 
 	uea_enters(INS_TO_USBDEV(sc));
 
-	INIT_WORK(&sc->task, uea_load_page, sc);
+	INIT_WORK(&sc->task, uea_load_page);
 	init_waitqueue_head(&sc->sync_q);
 	init_waitqueue_head(&sc->cmv_ack_wait);
 
